### PR TITLE
workflows: integrate inspire-matcher

### PR DESCRIPTION
**Pull request by jacquerie**
## Related Issue:
Closes https://github.com/inspirehep/inspire-next/issues/2527.

## Checklist:
- [x] I have all the information that I need (if not, move to `RFC` and look for it).
- [x] I linked the related issue(s) in the corresponding commit logs.
- [x] I wrote [good commit log messages](https://github.com/torvalds/subsurface-for-dirk/blob/5f15ad5a86ada3c5e574041a5f9d85235322dabb/README#L92-L119).
- [x] My code follows the code style of this project.
- [ ] I've added any new docs if API/utils methods were added.
- [x] I have updated the existing documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by jacquerie** [comment]
@ammirate Try rebasing your PR on top of this. I took care of making your life as little miserable as possible.

@david-caro This requires a reindexing of the Holding Pen to work properly!

File: inspirehep/modules/workflows/mappings/holdingpen/hep.json [hunk 1]
```diff
@@ -57,6 +57,43 @@
...
+                                "material": {
+                                    "type": "string"
+                                },
+                                "source": {
+                                    "type": "string"
+                                },
+                                "value": {
+                                    "fields": {
+                                        "raw": {
+                                            "index": "not_analyzed",
```
Comment (michamos) [inline inspirehep/modules/workflows/mappings/holdingpen/hep.json]: DOIs are case insensitive, so search in DOIs should be case-insensitive.

Comment (jacquerie) [reply inspirehep/modules/workflows/mappings/holdingpen/hep.json]: Right, in fact I made this mistake here too: https://github.com/inspirehep/inspire-next/blob/f64f0eee6a5a2c0eb996352f1f10371eb64cf530/inspirehep/modules/records/mappings/records/hep.json#L414

File: inspirehep/modules/workflows/tasks/matching.py [hunk 2]
```diff
@@ -268,58 +156,60 @@ def _previously_rejected(obj, eng):
...
+                        'match': 'dois.value',
+                        'search': 'metadata.dois.value.raw',
+                        'type': 'exact',
+                    },
+                ],
+            },
+        ],
+        'doc_type': 'hep',
+        'index': 'holdingpen-hep',
+    }
```
Comment (jacquerie) [inline inspirehep/modules/workflows/tasks/matching.py]: This should live in `inspirehep/modules/workflows/config.py`, but:
1. This file does not yet exist;
2. It's not yet loaded by `inspirehep/modules/workflows/ext.py`;
3. That file is a mess. Seriously, what's up with 2629743? https://github.com/inspirehep/inspire-next/blob/f64f0eee6a5a2c0eb996352f1f10371eb64cf530/inspirehep/modules/workflows/ext.py#L53-L54

File: inspirehep/modules/workflows/mappings/holdingpen/hep.json (hunk 1 above)
Comment (jacquerie) [reply inspirehep/modules/workflows/mappings/holdingpen/hep.json]: GitHub tried to be helpful here and attached this comment to the wrong line after I force pushed... I fixed this comment, and I also added a commit to amend the records mappings.

File: inspirehep/modules/workflows/tasks/matching.py [hunk 3]
```diff
@@ -203,20 +78,33 @@ def is_too_old(record, days_ago=5):
...
+        obj: a workflow object.
+        eng: a workflow engine.
+
+    Returns:
+        bool: ``True`` if the workflow object has a duplicate in the system
+        or ``extra_data`` contains a true value in ``match-found``, ``False``
+        otherwise.
+
+    """
+    if obj.extra_data.get('match-found'):
```
[the remark this answers is not in the document]
Comment (jacquerie) [reply inspirehep/modules/workflows/tasks/matching.py]: So you'd like to move this check in the workflow declaration?

File: inspirehep/modules/workflows/tasks/matching.py [hunk 4]
```diff
@@ -203,20 +78,33 @@ def is_too_old(record, days_ago=5):
...
-    else:
-        obj.log.warning("Remote match is deprecated.")
-        if match_legacy_inspire(obj, eng):
-            obj.log.info("Record already exists in INSPIRE.")
-            return True
+
+    matches = dedupe_list(match(obj.data))
+    record_ids = [el['_source']['control_number'] for el in matches]
+    if record_ids:
+        obj.extra_data['record_matches'] = record_ids
```
[the remark this answers is not in the document]
Comment (jacquerie) [reply inspirehep/modules/workflows/tasks/matching.py]: I like this :+1:

**Comment by jacquerie** [comment]
I think I'm going to merge that matcher, so that later when we merge the merger we will be able to merge the matches.

**Pull request merged**